### PR TITLE
update-codecgen.sh: support staging dirs in tsort logic

### LINKED ISSUE #48984: spurious changes in types.generated.go with hack/update-all.sh
I'm seeing changes show up in types.generated.go when running hack/update-all.sh that are unrelated to the changes I'm making, and cause CI to fail.

Example:
https://github.com/kubernetes/kubernetes/pull/45294/commits/4774babcacf1d8ae34e2300c8cc48f1ad6ca77bc#diff-84c1f4a54ef63a83dfee0975f6895d71
https://k8s-gubernator.appspot.com/pr/45294
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/45294/pull-kubernetes-verify/39502/

I ran with my PR rebased on latest master, ran `make clean` and `hack/godep-restore.sh` first, am running go1.8.3, etc

cc @caesarxuchao @sttts

### PR DESCRIPTION
**Pull request by sttts**
Fixes #48984.

### CODE REVIEW

**Comment by liggitt** [comment]
output of this doesn't look right to me:
```
$ hack/update-codecgen.sh | grep yes -B 1
DBG: does ./vendor/k8s.io/api/admission/v1alpha1/types.go depend on k8s.io/kubernetes/vendor/k8s.io/api/core/v1?
DBG: = yes
--
DBG: does ./vendor/k8s.io/api/admission/v1alpha1/types.go depend on k8s.io/kubernetes/vendor/k8s.io/api/apps/v1beta1?
DBG: = yes
--
DBG: does ./vendor/k8s.io/api/admission/v1alpha1/types.go depend on k8s.io/kubernetes/vendor/k8s.io/api/batch/v1?
DBG: = yes
$
```

I expected to see a lot more "yes" answers for things depending on core

**Comment by sttts** [comment]
We have meta now. Meta has no codecs. Maybe that's why?

**Comment by sttts** [comment]
But I agree, there are more, e.g. k8s.io/api/core/v1 is needed by apps/v1.

**Comment by sttts** [comment]
```
DBG: deps[./pkg/kubectl/testing/types.go]:
DBG: deps[./pkg/controller/garbagecollector/metaonly/types.go]:
DBG: deps[./pkg/apis/admission/types.go]:
DBG: deps[./federation/apis/federation/v1beta1/types.go]:   k8s.io/kubernetes/vendor/k8s.io/api/core/v1
DBG: deps[./cmd/libs/go2idl/client-gen/test_apis/testgroup/v1/types.go]:
DBG: deps[./vendor/k8s.io/metrics/pkg/apis/metrics/v1alpha1/types.go]:   k8s.io/kubernetes/vendor/k8s.io/api/core/v1
DBG: deps[./vendor/k8s.io/metrics/pkg/apis/metrics/types.go]:
DBG: deps[./vendor/k8s.io/metrics/pkg/apis/custom_metrics/v1alpha1/types.go]:   k8s.io/kubernetes/vendor/k8s.io/api/core/v1
DBG: deps[./vendor/k8s.io/metrics/pkg/apis/custom_metrics/types.go]:
DBG: deps[./vendor/k8s.io/apiserver/pkg/storage/testing/types.go]:
DBG: deps[./vendor/k8s.io/apiserver/pkg/endpoints/testing/types.go]:
DBG: deps[./vendor/k8s.io/apiserver/pkg/apis/example/v1/types.go]:
DBG: deps[./vendor/k8s.io/apiserver/pkg/apis/audit/v1alpha1/types.go]:
DBG: deps[./vendor/k8s.io/apimachinery/pkg/apis/testapigroup/v1/types.go]:
DBG: deps[./vendor/k8s.io/api/storage/v1beta1/types.go]:
DBG: deps[./vendor/k8s.io/api/rbac/v1beta1/types.go]:
DBG: deps[./vendor/k8s.io/api/rbac/v1alpha1/types.go]:
DBG: deps[./vendor/k8s.io/api/policy/v1beta1/types.go]:
DBG: deps[./vendor/k8s.io/api/networking/v1/types.go]:   k8s.io/kubernetes/vendor/k8s.io/api/core/v1
DBG: deps[./vendor/k8s.io/api/imagepolicy/v1alpha1/types.go]:
DBG: deps[./vendor/k8s.io/api/extensions/v1beta1/types.go]:   k8s.io/kubernetes/vendor/k8s.io/api/apps/v1beta1 k8s.io/kubernetes/vendor/k8s.io/api/core/v1
DBG: deps[./vendor/k8s.io/api/core/v1/types.go]:
DBG: deps[./vendor/k8s.io/api/certificates/v1beta1/types.go]:
DBG: deps[./vendor/k8s.io/api/batch/v2alpha1/types.go]:   k8s.io/kubernetes/vendor/k8s.io/api/batch/v1 k8s.io/kubernetes/vendor/k8s.io/api/core/v1
DBG: deps[./vendor/k8s.io/api/batch/v1/types.go]:   k8s.io/kubernetes/vendor/k8s.io/api/core/v1
DBG: deps[./vendor/k8s.io/api/autoscaling/v2alpha1/types.go]:   k8s.io/kubernetes/vendor/k8s.io/api/core/v1
DBG: deps[./vendor/k8s.io/api/autoscaling/v1/types.go]:   k8s.io/kubernetes/vendor/k8s.io/api/core/v1
DBG: deps[./vendor/k8s.io/api/authorization/v1beta1/types.go]:
DBG: deps[./vendor/k8s.io/api/authorization/v1/types.go]:
DBG: deps[./vendor/k8s.io/api/authentication/v1beta1/types.go]:
DBG: deps[./vendor/k8s.io/api/apps/v1beta1/types.go]:   k8s.io/kubernetes/vendor/k8s.io/api/core/v1
DBG: deps[./vendor/k8s.io/api/admissionregistration/v1alpha1/types.go]:
DBG: deps[./vendor/k8s.io/api/admission/v1alpha1/types.go]:
```

This looks about right.

**Comment by liggitt** [comment]
really? I'd expect all the external packages to depend on core

**Comment by sttts** [comment]
A handful looked right. But vendor/k8s.io/api/admission/v1alpha1/types.go depends on vendor/k8s.io/api/authentication/v1. This is not reflected in the output.

**Comment by sttts** [comment]
False alarm (or an unrelated issue): k8s.io/api/authentication/v1 has no codec (why?).

File: hack/update-codecgen.sh [hunk 1]
```diff
@@ -121,7 +113,8 @@ function depends {
 function tsort {
   visited[$1]=true
   local j=0
-  for (( j=0; j<number; j++ )); do
```
Comment (caesarxuchao) [inline hack/update-codecgen.sh]: Why is this change? It looks like `number` and `generated_files` don't change.

**Review by caesarxuchao — COMMENTED** [review]
A few nits. I'll run the script to verify it.

File: hack/update-codecgen.sh [hunk 2]
```diff
@@ -43,36 +40,35 @@ generated_files=($(
...
+dep_packages=($(
+    for f in "${generated_files[@]}"; do
+        package "${f}"
+    done | LC_ALL=C sort -u
 ))
-
-# We only work for deps within this prefix.
-my_prefix="k8s.io/kubernetes"
+###for f in $(echo "${dep_packages[@]}"); do
+###    echo "dep_packages:   $f"
```
Comment (sttts) [inline hack/update-codecgen.sh]: done

File: hack/update-codecgen.sh [hunk 3]
```diff
@@ -81,31 +77,27 @@ my_prefix="k8s.io/kubernetes"
...
-  deps[${i}]=$(go list -f '{{range .Deps}}{{.}}{{"\n"}}{{end}}' ${file} | grep "^${my_prefix}")
+  deps[${i}]=$(go list -f '{{range .Deps}}{{.}}{{"\n"}}{{end}}' ${file} | grep -Eow "^\\(${dep_packages_grep_pattern}\\)\$" || true)
+  ###echo "DBG: deps[$file]:   ${deps[$i]}"
 done
-###echo "DBG: found $number generated files"
-###for f in $(echo "${generated_files[@]}" | LC_ALL=C sort); do
-###    echo "DBG:   $f"
-###done
 
 # NOTE: depends function assumes that the whole repository is under
```
Comment (sttts) [inline hack/update-codecgen.sh]: done

File: hack/update-codecgen.sh (hunk 1 above)
Comment (sttts) [inline hack/update-codecgen.sh]: fixed

File: hack/update-codecgen.sh [hunk 4]
```diff
@@ -81,32 +77,28 @@ my_prefix="k8s.io/kubernetes"
...
-###    echo "DBG:   $f"
-###done
 
-# NOTE: depends function assumes that the whole repository is under
-# $my_prefix - it will NOT work if that is not true.
 function depends {
-  rhs="$(dirname ${generated_files[$2]/#./${my_prefix}})"
-  ###echo "DBG: does ${file} depend on ${rhs}?"
+  rhs="$(package ${generated_files[$2]})"
+  ####echo "DBG: does ${file} depend on ${rhs}?"
```
Comment (caesarxuchao) [inline hack/update-codecgen.sh]: It should be generated_files[$1], not ${file} here.

**Comment by caesarxuchao** [comment]
If a dependent package (assuming "A") happens to be visited earlier then its importer (assuming "B"), then the tsort skips testing if "B" depends on "A" since "A" is visited. That explains why there are so few "yes" in the output.

**Comment by caesarxuchao** [comment]
The sorted order looks right to me. @liggitt @sttts could you double check?

```
./pkg/kubectl/testing/types.generated.go
./pkg/controller/garbagecollector/metaonly/types.generated.go
./pkg/apis/admission/types.generated.go      // we should remove this file
./vendor/k8s.io/api/core/v1/types.generated.go
./federation/apis/federation/v1beta1/types.generated.go
./cmd/libs/go2idl/client-gen/test_apis/testgroup/v1/types.generated.go
./vendor/k8s.io/metrics/pkg/apis/metrics/v1alpha1/types.generated.go
./vendor/k8s.io/metrics/pkg/apis/metrics/types.generated.go
./vendor/k8s.io/metrics/pkg/apis/custom_metrics/v1alpha1/types.generated.go
./vendor/k8s.io/metrics/pkg/apis/custom_metrics/types.generated.go
./vendor/k8s.io/apiserver/pkg/storage/testing/types.generated.go
./vendor/k8s.io/apiserver/pkg/endpoints/testing/types.generated.go
./vendor/k8s.io/apiserver/pkg/apis/example/v1/types.generated.go
./vendor/k8s.io/api/authentication/v1/types.generated.go
./vendor/k8s.io/apiserver/pkg/apis/audit/v1alpha1/types.generated.go
./vendor/k8s.io/apimachinery/pkg/apis/testapigroup/v1/types.generated.go
./vendor/k8s.io/api/storage/v1beta1/types.generated.go
./vendor/k8s.io/api/storage/v1/types.generated.go
./vendor/k8s.io/api/settings/v1alpha1/types.generated.go
./vendor/k8s.io/api/rbac/v1beta1/types.generated.go
./vendor/k8s.io/api/rbac/v1alpha1/types.generated.go
./vendor/k8s.io/api/policy/v1beta1/types.generated.go
./vendor/k8s.io/api/networking/v1/types.generated.go
./vendor/k8s.io/api/imagepolicy/v1alpha1/types.generated.go
./vendor/k8s.io/api/apps/v1beta1/types.generated.go
./vendor/k8s.io/api/extensions/v1beta1/types.generated.go
./vendor/k8s.io/api/certificates/v1beta1/types.generated.go
./vendor/k8s.io/api/batch/v1/types.generated.go
./vendor/k8s.io/api/batch/v2alpha1/types.generated.go
./vendor/k8s.io/api/autoscaling/v2alpha1/types.generated.go
./vendor/k8s.io/api/autoscaling/v1/types.generated.go
./vendor/k8s.io/api/authorization/v1beta1/types.generated.go
./vendor/k8s.io/api/authorization/v1/types.generated.go
./vendor/k8s.io/api/authentication/v1beta1/types.generated.go
./vendor/k8s.io/api/admissionregistration/v1alpha1/types.generated.go
./vendor/k8s.io/api/admission/v1alpha1/types.generated.go
```

**Comment by liggitt** [comment]
godep check flake, /retest

**Comment by liggitt** [comment]
> If a dependent package (assuming A) happens to be visited earlier then its importer (assuming B), then the tsort skips testing if B depends on A since A is visited. That explains why there are so few "yes" in the output.

ah

I scanned the order briefly and nothing jumped out at me as being out of order.

Should these be removed as well since they're internal types:
./vendor/k8s.io/metrics/pkg/apis/metrics/types.generated.go
./vendor/k8s.io/metrics/pkg/apis/custom_metrics/types.generated.go

**Comment by caesarxuchao** [comment]
Deleted internal types.generated.go in https://github.com/kubernetes/kubernetes/pull/49240.

File: hack/update-codecgen.sh (hunk 4 above)
Comment (sttts) [inline hack/update-codecgen.sh]: fixed

**Comment by liggitt** [comment]
picked up an extra commit, LGTM otherwise

**Comment by sttts** [comment]
Split out the missing codecs change into https://github.com/kubernetes/kubernetes/pull/49302, to get this in faster.

**Pull request merged**